sparse/unit_test/Test_Sparse_spmv_bsr.hpp

```diff
@@ -14,563 +14,636 @@
 //
 //@HEADER
 
+/*! \file Test_Sparse_spmv_bsr.hpp
+
+  Test the following 768 combos for at least a few matcies.
+
+  Algorithms              Alpha     Beta     Block Sizes    Modes
+  (none)                  0         0        1              N
+  native              x   1      x  1     x  2           x  T
+  experimental_bsr_tc     -1        -1       5              C
+                          3.7       -1.5     9              H
+
+  There are also a subset of tests on larger matrices
+*/
+
 #include <algorithm>
+#include <iostream>
+#include <stdexcept>
+
 #include <gtest/gtest.h>
 #include <Kokkos_Core.hpp>
-#include <stdexcept>
-#include "KokkosSparse_spmv.hpp"
-#include "KokkosSparse_BsrMatrix.hpp"
-#include "KokkosSparse_CrsMatrix.hpp"
 
 #include <KokkosKernels_TestUtils.hpp>
 #include <KokkosKernels_Test_Structured_Matrix.hpp>
 #include <KokkosKernels_IOUtils.hpp>
 #include <KokkosKernels_Utils.hpp>
-
 #include "KokkosKernels_Controls.hpp"
 #include "KokkosKernels_default_types.hpp"
 
-typedef Kokkos::complex<double> kokkos_complex_double;
-typedef Kokkos::complex<float> kokkos_complex_float;
+#include "KokkosSparse_spmv.hpp"
+#include "KokkosSparse_BsrMatrix.hpp"
+#include "KokkosSparse_CrsMatrix.hpp"
+#include "KokkosSparse_crs_to_bsr_impl.hpp"
+#include "KokkosSparse_bsr_to_crs_impl.hpp"
+#include "KokkosSparse_Utils.hpp"
+
+using kokkos_complex_double = Kokkos::complex<double>;
+using kokkos_complex_float  = Kokkos::complex<float>;
 
-namespace Test_Bsr {
+namespace Test_Spmv_Bsr {
 
-/// Random generator
-template <typename Scalar>
-inline Scalar random() {
-  auto const max = static_cast<Scalar>(RAND_MAX) + static_cast<Scalar>(1);
-  return static_cast<Scalar>(std::rand()) / max;
+/*! \brief Maximum value used to fill A */
+template <typename T>
+constexpr T max_a() {
+  return T(10.0);
 }
 
-template <typename Scalar>
-inline void set_random_value(Scalar &v) {
-  v = random<Scalar>();
+/*! \brief Maximum value used to fill X */
+template <typename T>
+constexpr T max_x() {
+  return T(10.0);
 }
 
-template <typename Scalar>
-inline void set_random_value(Kokkos::complex<Scalar> &v) {
-  Scalar vre = random<Scalar>();
-  Scalar vim = random<Scalar>();
-  v          = Kokkos::complex<Scalar>(vre, vim);
+/*! \brief Maximum value used to fill Y */
+template <typename T>
+constexpr T max_y() {
+  return T(10.0);
```

Comment:
It looks like A, x, y will all be generated with imaginary part 0 if T is a complex type. You could use the getRandomBounds utility function:
```
T discard, maxVal;
KokkosKernels::Impl::getRandomBounds(10.0, discard, maxVal);
return maxVal;
```
to return 10 for real T, and 10+10i for complex T. "discard" is -10 - 10i in this case, but since you're sampling from 0...max it's not needed.

Reply:
Very nice, thanks!